src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+
+            return;
+        }
+
+        IList<ExpressionStatementSyntax> fieldInitializersInConstructors = fieldInfo.FieldInitializersInConstructors;
+
+        // If the field is initialized in the declaration and in the constructors,
+        // we check if the initializations are diverging.
+        // Since we already checked if the field should not initialize in the declaration,
+        // we know all initializations in the constructors are the same.
+        // So we only need to check if the initializer is diverging from the constructor initializations.
+        if (isInitializerPresent
+            && fieldInitializersInConstructors.Count != 0
+            && HasDivergingInitializations(fieldInitializersInConstructors[0], initializer))
+        {
+            context.ReportDiagnostic(field.GetLocation().CreateDiagnostic(RuleExceptionShouldNotInitializeInDeclaration));
+            return;
+        }
+
+        // We know the initializers in the constructors are the same, so we report diagnostics (general rule) on them.
+        ReportDiagnosticsForInitializersInConstructors(context, fieldInitializersInConstructors);
+    }
+
+    /// <summary>
+    /// Checks if the initializers in the constructors are diverging from the initializer in the declaration.
+    /// </summary>
+    /// <param name="initializerInConstructor">A <see cref="ExpressionStatementSyntax"/> from the constructor.</param>
+    /// <param name="initializerInDeclaration">The initializer from the declaration.</param>
+    /// <returns>A bool on whether the constructor initializers diverge from the declaration.</returns>
+    private static bool HasDivergingInitializations(ExpressionStatementSyntax initializerInConstructor, EqualsValueClauseSyntax initializerInDeclaration)
+    {
+        return !((AssignmentExpressionSyntax)initializerInConstructor.Expression).Right.IsEquivalentTo(initializerInDeclaration.Value);
+    }
+
+    /// <summary>
+    /// Reports diagnostics for initializers in constructors <see cref="GeneralRule"/>.
+    /// </summary>
+    /// <param name="context">The analysis context.</param>
+    /// <param name="initializers">The list of constructor field initializers.</param>
+    private static void ReportDiagnosticsForInitializersInConstructors(SyntaxNodeAnalysisContext context, IList<ExpressionStatementSyntax> initializers)
+    {
+        for (int i = 0; i < initializers.Count; i++)
+        {
+            context.ReportDiagnostic(initializers[i].GetLocation().CreateDiagnostic(GeneralRule));
+        }
+    }
+
+    /// <summary>
+    /// A record to track field initialization information in constructors.
+    /// </summary>
+    private sealed record FieldInitializationInfo
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FieldInitializationInfo"/> class.
+        /// </summary>
+        /// <param name="name">The field name.</param>
+        public FieldInitializationInfo(
+            string name)
```

Comment:
No need to wrap here, line isn't long enough

Reply:
Removing constructor since I don't need it